File: kompartment/src/ode/solvers/dormand-prince.js
/**
 * The explicit Runge-Kutta (4,5) pair of Dormand and Prince.
 *
 * The tableau is from J. R. Dormand and P. J. Prince, "A family of embedded
 * Runge-Kutta formulae", J. Comp. Appl. Math. 6(1) 19-26 (1980): seven
 * stages, the last evaluated at the new point so that it is the first stage
 * of the next step, and a fifth-order solution with a fourth-order error
 * estimate. The dense output is the quartic interpolant that comes free with
 * the pair, as tabulated by Hairer, Nørsett and Wanner, *Solving Ordinary
 * Differential Equations I*, section II.6.
 *
 * Everything around the step -- step-size control, events, output, the
 * constraint -- is the shared driver in ./onestep.js. Non-negativity here
 * participates only in the error test: an explicit pair cannot hold a state
 * on its bound, since its stages straddle the kink and disagree by the whole
 * jump, so a constraint that binds is something this method reports rather
 * than carries.
 */

import { integrate, SolverError, nonFiniteError, STALL_WINDOW_STEPS, STALL_SPAN_FRACTION } from './onestep.js';

export { SolverError, nonFiniteError, STALL_WINDOW_STEPS, STALL_SPAN_FRACTION };

/** Nodes c_2..c_7 of the tableau. */
const NODES = [1 / 5, 3 / 10, 4 / 5, 8 / 9, 1, 1];

/** Coefficients a_ij of stages 2..7 on the stages before them. */
const STAGE = [
	[1 / 5],
	[3 / 40, 9 / 40],
	[44 / 45, -56 / 15, 32 / 9],
	[19372 / 6561, -25360 / 2187, 64448 / 6561, -212 / 729],
	[9017 / 3168, -355 / 33, 46732 / 5247, 49 / 176, -5103 / 18656],
	[35 / 384, 0, 500 / 1113, 125 / 192, -2187 / 6784, 11 / 84],
];

/** The difference between the fifth- and fourth-order solutions, per stage. */
const ERROR = [71 / 57600, 0, -71 / 16695, 71 / 1920, -17253 / 339200, 22 / 525, -1 / 40];

/** The free interpolant: per stage, the coefficients of s, s², s³ and s⁴. */
const DENSE = [
	[1, -183 / 64, 37 / 12, -145 / 128],
	[0, 0, 0, 0],
	[0, 1500 / 371, -1000 / 159, 1000 / 371],
	[0, -125 / 32, 125 / 12, -375 / 64],
	[0, 9477 / 3392, -729 / 106, 25515 / 6784],
	[0, -11 / 7, 11 / 3, -55 / 28],
	[0, 3 / 2, -4, 5 / 2],
];

export const dormandPrinceMethod = {
	id: 'dp45',
	order: 4,
	defaultMaxSteps: 1e7,
	holdsAtZero: false,
	snapsToConstraint: false,

	stepBudgetMessage: (maxSteps, t) => (
		`Exceeded ${maxSteps} steps at t=${t}; the system may be stiff `
		+ '-- try a stiff solver (variableOrder or rosenbrock23).'
	),
	floorMessage: (t, hmin, worst) => (
		`Unable to meet integration tolerances at t=${t} without reducing the `
		+ `step below the smallest allowed (${hmin}). State ${worst} is the `
		+ 'worst offender; the system is probably stiff -- try a stiff '
		+ 'solver (variableOrder or rosenbrock23).'
	),
	stallMessage: (t, window, spanCovered) => (
		`The solver stopped making progress at t=${t}: ${window} steps `
		+ `advanced the clock by less than ${spanCovered}. Either the system is stiff `
		+ '-- try a stiff solver (variableOrder or rosenbrock23) -- or a discontinuous rate '
		+ 'is holding a state against zero.'
	),

	setUp({ neq, rhs }) {
		const k = new Array(7);
		for (let i = 0; i < 7; i++) k[i] = new Float64Array(neq);
		const stage = new Float64Array(neq);
		// The step the interpolant belongs to.
		let tFrom = 0;
		let hFrom = 0;
		let yFrom = null;

		const combine = (y, h, row, out) => {
			for (let i = 0; i < neq; i++) {
				let acc = 0;
				for (let j = 0; j < row.length; j++) if (row[j] !== 0) acc += row[j] * k[j][i];
				out[i] = y[i] + h * acc;
			}
			return out;
		};

		return {
			start(t, y) {
				rhs(t, y, k[0]);
				return 1;
			},
			derivativeAtStart() {
				return k[0];
			},
			attempt(t, y, h, tnew, ynew, err) {
				for (let s = 0; s < 5; s++) {
					rhs(t + h * NODES[s], combine(y, h, STAGE[s], stage), k[s + 1]);
				}
				combine(y, h, STAGE[5], ynew);
				rhs(tnew, ynew, k[6]);
				for (let i = 0; i < neq; i++) {
					let acc = 0;
					for (let j = 0; j < 7; j++) if (ERROR[j] !== 0) acc += ERROR[j] * k[j][i];
					err[i] = acc;
				}
				tFrom = t;
				hFrom = h;
				yFrom = y;
				return { fevals: 6, scale: 1 };
			},
			denseAt(tq, out) {
				const s = (tq - tFrom) / hFrom;
				const s2 = s * s;
				const s3 = s2 * s;
				const s4 = s3 * s;
				for (let i = 0; i < neq; i++) {
					let acc = 0;
					for (let j = 0; j < 7; j++) {
						const c = DENSE[j];
						const w = c[0] * s + c[1] * s2 + c[2] * s3 + c[3] * s4;
						if (w !== 0) acc += w * k[j][i];
					}
					out[i] = yFrom[i] + hFrom * acc;
				}
				return out;
			},
			accept(tnew, ynew, reprojected) {
				// The last stage is the first of the next step -- unless the
				// state it was evaluated at has since been projected.
				let spent = 0;
				if (reprojected) { rhs(tnew, ynew, k[6]); spent = 1; }
				k[0].set(k[6]);
				return spent;
			},
			restart() {
				return 0;
			},
			stats() {
				return {};
			},
		};
	},
};

/**
 * @param {(t: number, y: Float64Array, out: Float64Array) => Float64Array} f
 * @param {number[]|Float64Array} tspan  output grid; first and last bound the run
 * @param {Float64Array} y0
 * @param {object} opts  see ./onestep.js
 * @returns {{ t: Float64Array, y: Float64Array[], stats: object }}
 */
export function dormandPrince(f, tspan, y0, opts = {}) {
	return integrate(dormandPrinceMethod, f, tspan, y0, opts);
}

File: kompartment/python/kompartment/engine/solvers/dormand_prince.py
"""Dormand-Prince (4,5): the explicit method (``src/ode/solvers/dormand-prince.js``).

Seven stages with the first-same-as-last property, error controlled on the
fourth-order solution, and the continuous extension of Hairer-Norsett-Wanner
for output between steps.
"""

from __future__ import annotations

from typing import Any, Dict, Tuple

import numpy as np

from .onestep import integrate

NODES = (1 / 5, 3 / 10, 4 / 5, 8 / 9, 1.0, 1.0)
STAGE = (
    (1 / 5,),
    (3 / 40, 9 / 40),
    (44 / 45, -56 / 15, 32 / 9),
    (19372 / 6561, -25360 / 2187, 64448 / 6561, -212 / 729),
    (9017 / 3168, -355 / 33, 46732 / 5247, 49 / 176, -5103 / 18656),
    (35 / 384, 0.0, 500 / 1113, 125 / 192, -2187 / 6784, 11 / 84),
)
ERROR = (71 / 57600, 0.0, -71 / 16695, 71 / 1920, -17253 / 339200, 22 / 525, -1 / 40)
DENSE = (
    (1.0, -183 / 64, 37 / 12, -145 / 128),
    (0.0, 0.0, 0.0, 0.0),
    (0.0, 1500 / 371, -1000 / 159, 1000 / 371),
    (0.0, -125 / 32, 125 / 12, -375 / 64),
    (0.0, 9477 / 3392, -729 / 106, 25515 / 6784),
    (0.0, -11 / 7, 11 / 3, -55 / 28),
    (0.0, 3 / 2, -4.0, 5 / 2),
)


class _Stepper:
    def __init__(self, neq: int, rhs: Any) -> None:
        self.neq = neq
        self.rhs = rhs
        self.k = [np.zeros(neq) for _ in range(7)]

    def _combine(self, y: np.ndarray, h: float, row: Tuple[float, ...]) -> np.ndarray:
        acc = np.zeros(self.neq)
        for j, c in enumerate(row):
            if c != 0:
                acc = acc + c * self.k[j]
        return y + h * acc

    def start(self, t: float, y: np.ndarray) -> int:
        self.k[0] = np.asarray(self.rhs(t, y), dtype=float)
        return 1

    def derivative_at_start(self) -> np.ndarray:
        return self.k[0]

    def attempt(self, t: float, y: np.ndarray, h: float, tnew: float, ynew: np.ndarray,
                err: np.ndarray) -> Tuple[int, float]:
        for s in range(5):
            self.k[s + 1] = np.asarray(self.rhs(t + h * NODES[s], self._combine(y, h, STAGE[s])), dtype=float)
        ynew[:] = self._combine(y, h, STAGE[5])
        self.k[6] = np.asarray(self.rhs(tnew, ynew), dtype=float)
        acc = np.zeros(self.neq)
        for j in range(7):
            if ERROR[j] != 0:
                acc = acc + ERROR[j] * self.k[j]
        err[:] = acc
        self.t_from, self.h_from, self.y_from = t, h, y.copy()
        return 6, 1.0

    def dense_at(self, tq: float) -> np.ndarray:
        s = (tq - self.t_from) / self.h_from
        s2 = s * s
        s3 = s2 * s
        s4 = s3 * s
        acc = np.zeros(self.neq)
        for j in range(7):
            c = DENSE[j]
            w = c[0] * s + c[1] * s2 + c[2] * s3 + c[3] * s4
            if w != 0:
                acc = acc + w * self.k[j]
        return self.y_from + self.h_from * acc

    def accept(self, tnew: float, ynew: np.ndarray, reprojected: bool) -> int:
        spent = 0
        if reprojected:
            self.k[6] = np.asarray(self.rhs(tnew, ynew), dtype=float)
            spent = 1
        self.k[0] = self.k[6].copy()
        return spent

    def restart(self, t: float, y: np.ndarray) -> int:
        return 0

    def stats(self) -> Dict[str, Any]:
        return {}


class _Method:
    id = 'dp45'
    order = 4
    default_max_steps = 1e7
    holds_at_zero = False
    snaps_to_constraint = False

    @staticmethod
    def step_budget_message(max_steps: float, t: float) -> str:
        return (f'Exceeded {int(max_steps)} steps at t={t}; the system may be stiff -- try a stiff solver '
                '(variableOrder or rosenbrock23).')

    @staticmethod
    def floor_message(t: float, hmin: float, worst: int) -> str:
        return (f'Unable to meet integration tolerances at t={t} without reducing the step below the smallest allowed '
                f'({hmin}). State {worst} is the worst offender; the system is probably stiff -- try a stiff solver '
                '(variableOrder or rosenbrock23).')

    @staticmethod
    def stall_message(t: float, window: int, covered: float) -> str:
        return (f'The solver stopped making progress at t={t}: {window} steps advanced the clock by less than '
                f'{covered}. Either the system is stiff -- try a stiff solver (variableOrder or rosenbrock23) -- or a '
                'discontinuous rate is holding a state against zero.')

    @staticmethod
    def set_up(neq: int, rhs: Any, f: Any, threshold: np.ndarray, atol: np.ndarray, rtol: float,
               opts: Dict[str, Any]) -> _Stepper:
        return _Stepper(neq, rhs)


def dormand_prince(f: Any, tspan: Any, y0: np.ndarray, opts: Dict[str, Any]) -> Dict[str, Any]:
    return integrate(_Method, f, tspan, y0, opts)
